File: 슈퍼컴퓨팅문제/KSC2016_ProblemSet_n_Solution.tar/KSC2016_ProblemSet_n_Solution/Prob1.tar/Prob1/Prob1/Code/c_serial/kmeans.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "kmeans.h"

void assignment_step(const PPOINT kmeans, const PPOINT pt, int *kindex)
{
	double d1, d2;

	for(int i=0; i<N_PT; i++) {
		kindex[i] = 0;
		d1 = distance(pt[i], kmeans[0]);
		for(int j=1; j<N_K; j++) {
			d2 = distance(pt[i], kmeans[j]);
			if( d1 > d2 ) {
				d1 = d2;
				kindex[i] = j;
			}
		}
	}
}

void update_step(PPOINT kmeans, const PPOINT pt, const int *kindex)
{
	int i, idx;
	int num_pt[N_K];

	for(i=0; i<N_K; i++) {
		kmeans[i].x = kmeans[i].y = 0.0;
		num_pt[i] = 0;
	}
	for(i=0; i<N_PT; i++) {
		idx = kindex[i];
		kmeans[idx].x += pt[i].x;
		kmeans[idx].y += pt[i].y;
		num_pt[idx]++;
	}

	for(i=0; i<N_K; i++) {
		kmeans[i].x /= num_pt[i];
		kmeans[i].y /= num_pt[i];
	}
}

File: 슈퍼컴퓨팅문제/KSC2016_ProblemSet_n_Solution.tar/KSC2016_ProblemSet_n_Solution/Prob1.tar/Prob1/Prob1/Code/f_serial/kmeans.f90
module kmeans_mod
    use util_mod
    implicit none

contains

subroutine assignment_step(kmeans, pt, kindex)
    implicit none
    type(POINT), intent(in) :: kmeans(N_K), pt(N_PT)
    integer, intent(out) :: kindex(N_PT)
    integer :: i,j
    real(8) :: d1, d2

    do i=1, N_PT
        kindex(i) = 1
        d1 = distance(pt(i), kmeans(1))
        do j=2,N_K
            d2 = distance(pt(i), kmeans(j))
            if( d1 > d2 ) then
                d1 = d2
                kindex(i) = j
            end if
        end do
    end do
end subroutine

subroutine update_step(kmeans, pt, kindex)
    implicit none
    integer, intent(in) :: kindex(N_PT)
    type(POINT), intent(in) :: pt(N_PT)
    type(POINT), intent(out) :: kmeans(N_K)
    integer :: i, idx, num_pt(N_K), ierr

    kmeans%x = 0.0
    kmeans%y = 0.0
    num_pt = 0

    do i=1, N_PT
        idx = kindex(i)
        kmeans(idx)%x = kmeans(idx)%x + pt(i)%x
        kmeans(idx)%y = kmeans(idx)%y + pt(i)%y
        num_pt(idx) = num_pt(idx) + 1
    end do
	
    kmeans%x = kmeans%x / num_pt
    kmeans%y = kmeans%y / num_pt
end subroutine

end module
